File: kite/graph-core/src/main/kotlin/com/kite/di/graph/Decisions.kt
package com.kite.di.graph

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * `decisions.json` — the pending-decisions artifact behind board decision cards
 *. Written by the KSP processor next to graph.json on every
 * debug compilation: with the decisions blocking inference when the build
 * fails, with an empty [DecisionsFile.pending] on success (which clears the
 * cards). A development-only artifact, never packaged — same rules as
 * graph.json.
 */

/** One way to answer a decision — a click on this candidate writes [insert]. */
@Serializable
data class DecisionCandidate(
    val fqn: String,
    val displayName: String,
    /** Where the candidate is declared (repo-root-relative). */
    val file: String,
    val line: Int,
    /** The exact rules line this candidate resolves the decision with. */
    val insert: String,
)

@Serializable
data class PendingDecision(
    /** Stable id, e.g. `"bind:com.app.data.UserRepository"`. */
    val id: String,
    /** `"bind"` today; `"root"` / `"scope"` reserved for later cards. */
    val kind: String,
    /** Card headline, e.g. `"UserRepository has 2 implementations"`. */
    val title: String,
    /** The type the decision is about (FQN). */
    val subject: String,
    /** Human-readable consumer sites that force the decision. */
    val consumers: List<String> = emptyList(),
    val candidates: List<DecisionCandidate> = emptyList(),
)

@Serializable
data class DecisionsFile(
    val module: String,
    /** Repo-root-relative path of the module's rules holder file, when one exists. */
    val rulesFile: String? = null,
    /** Simple name of the holder object inside [rulesFile]. */
    val rulesObject: String? = null,
    /** Where the board server should create a holder when the module has none. */
    val suggestedFile: String? = null,
    val suggestedPackage: String? = null,
    val pending: List<PendingDecision> = emptyList(),
)

object DecisionsJson {
    private val json: Json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        explicitNulls = false
    }
    private val prettyJson: Json = Json(json) { prettyPrint = true }

    fun encodePretty(file: DecisionsFile): String =
        prettyJson.encodeToString(DecisionsFile.serializer(), file)

    fun decode(text: String): DecisionsFile =
        json.decodeFromString(DecisionsFile.serializer(), text)
}

File: webboard/src/model/decisions.ts
/**
 * TypeScript mirror of decisions.json produced by the KSP
 * processor next to graph.json: the pending decisions blocking inference,
 * rendered by the board as clickable cards. Kotlin side:
 * `kite/graph-core` Decisions.kt.
 */

export interface DecisionCandidate {
  fqn: string;
  displayName: string;
  /** Where the candidate is declared (repo-root-relative). */
  file: string;
  line: number;
  /** The exact rules line a click writes. */
  insert: string;
}

export interface PendingDecision {
  /** Stable id, e.g. `bind:com.app.data.UserRepository`. */
  id: string;
  /** 'bind' today; 'root' / 'scope' reserved. */
  kind: string;
  title: string;
  subject: string;
  consumers: string[];
  candidates: DecisionCandidate[];
}

export interface DecisionsFile {
  module: string;
  rulesFile?: string | null;
  rulesObject?: string | null;
  suggestedFile?: string | null;
  suggestedPackage?: string | null;
  pending: PendingDecision[];
}

export function emptyDecisions(): DecisionsFile {
  return { module: '', pending: [] };
}

/** Light structural validation of an untrusted decisions.json payload. */
export function parseDecisions(raw: unknown): DecisionsFile {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) return emptyDecisions();
  const rec = raw as Record<string, unknown>;
  if (!Array.isArray(rec['pending'])) return emptyDecisions();
  const pending: PendingDecision[] = [];
  for (const p of rec['pending'] as unknown[]) {
    if (typeof p !== 'object' || p === null) continue;
    const d = p as Record<string, unknown>;
    if (typeof d['id'] !== 'string' || typeof d['title'] !== 'string' || !Array.isArray(d['candidates'])) continue;
    pending.push(d as unknown as PendingDecision);
  }
  return { ...(rec as unknown as DecisionsFile), pending };
}
